### DISCUSSION TITLE
Inquiry on SPINUP_LOOPS Parameter Settings and Parallel Computing Tutorial

### DISCUSSION THREAD

**Discussion by ZRYings** [NCAR/hrldas#118] [Q&A] [answered]
Hi, I am planning to conduct simulations over a large area spanning 55 years. (1) Before the official simulation, I aim to conduct a spinup using the first 5 years of forcing data as a baseline. Could you kindly advise on how the SPINUP_LOOPS parameter should be set to achieve this? 
(2) In addition, would it be possible to provide me a tutorial on setting up parallel computing as mentioned in this link (https://github.com/NCAR/hrldas/blob/master/tutorial/Note2_2D_NLDAS2_domain.ipynb)?

Thank you very much for your time!

**Comment by cenlinhe** [comment]
Hi, to answer your question:
(1) based on my understanding, you plan to use the first 5-year of your forcing to run a 5-year spin-up, right? Are you going to discard the first 5-year and only use the later 50 years' data? If so, you do not need to set up SPINUP_LOOPS (i.e., put it as the default value, SPINUP_LOOPS = 0) and just run a continuous 55-year simulation and not using the first 5-year data. Did I mis-understand what you want to do?
(2) The model already had parallel computing capabilities. To activate it, it really depends on your own machine setup (e.g., use parallel compiler such as mpi instead of ifort). Once you choose the parallel mode for your machine environment and compiler, then when you select the model during ./configure setup, choose the compiler option consistent with your machine environment (e.g., "Linux ifort compiler MPI" if using intel parallel compiler). Then edit the generated user_build_options file for netcdf and Jasper software path. And then compile the code. You can follow this instruction: https://github.com/NCAR/hrldas/blob/master/tutorial/Note0_Download_Compile.ipynb

**Reply by ZRYings** [reply]
Hi, thank you very much for your reply. (1) I apologize for the confusion regarding the SPINUP_LOOPS setting. I would like to use simulation data for the entire 55 years, but I found it is very time-consuming if setting SPINUP_LOOPS to a number _n_, cause the model runs the 55-year data (_n_+1) times. So, I am thinking to first conduct a spin-up using just the initial 5-year data and then run the simulation for the remaining 50 years. Would it be okay? If so, should I perform two separate runs? For example, in the first run, set the simulation duration to kday=365*5 and set SPINUP_LOOPS to the desired number, saving the last day as a restart file. For the second run, use this restart file and set SPINUP_LOOPS to 0 to simulate the remaining 50 years. I hope I have made myself clear, and I look forward to your advice.

Secondly, we have set up an MPI parallel environment according to the tutorial. But I saw section 4.2 in this link (https://github.com/NCAR/hrldas/blob/master/tutorial/Note2_2D_NLDAS2_domain.ipynb) mentioned that 'for larger domains and longer simulation times, it's encouraged to submit jobs to a supercomputer, e.g., Cheyenne, using a job submission script called runhrldas.csh and submitted via _qsub runhrldas.csh_'. Could you please share the script of the runhrldas.csh?

Thank you and looking forward to your reply.

**Reply by cenlinhe** [reply, marked as answer]
(1) How many total years do you want to do your spin-up? If you only need 5-year spin-up, then you do not need two separate runs. You can do a continuous 55-year simulation (SPINUP_LOOPS = 0), and then only analyze your recent 50-year results. If you need, for example, 15-year spin up, then you need two separate runs, the first run using the first 5-year forcing to run the model by setting SPINUP_LOOPS = 3, and use the produced restart file for the second 50-year run. You can set restart file output frequency to be daily.
(2) Our runhrldas.csh only works in our NCAR supercomputer environment. Are you running the model in NCAR supercomputer? It has to be specific about your own machine. Different machines use different job submission systems. You may want to check with your IT about the job submission system.

**Reply by ZRYings** [reply]
Hi, thank you very much for your detailed guidance.

**Reply by sudarsan3010** [reply]
Hi @cenlinhe,

For my study area, I plan to run the spinup several times for one particular initial year (say 2008) until the model gets stabilized.  My query is that if I give the SPINUP_LOOPS = 20, will all the 20 simulation outputs be saved separately or will it be overwritten by the last run's output?

I am asking this because I need to check with the previous year's output and compute the percentage change.

**Reply by cenlinhe** [reply]
Yes, the output files will all have different names for each spinup loop with a suffix of loop number "_loop0001 ", like yyyymmddhh.LDOUT_DOMAIN1.loop*. However, the restart file will be overwritten after each loop to get updated.

**Reply by sudarsan3010** [reply]
Dear @cenlinhe, 

Thank you so much for your quick reply. Kindly address the following querries as well.

I have created forcing for 15 years from 2008 to 2023 and the variables are present in LDASIN directory with a setup file 'HRLDAS_setup_2008010100_d1'.

1. I am planning to spinup the model over the year of 2008 for 'n' times, and start the simulation from 2009. In this case, the setup file can be same as that of 'HRLDAS_setup_2008010100_d1' or should I need a separate setup file for 2009?

2. Other than changing the below line, what are the other changes is to be made to run the model in 'Restart' mode.
! RESTART_FILENAME_REQUESTED = "/glade/scratch/cenlinhe/test/hrldas_output/RESTART.2000010100_DOMAIN1"

**Reply by RAbolafiaRosenzweig** [reply]
![image](https://github.com/NCAR/hrldas/assets/36647844/a15f459c-618f-4fce-a408-e5f59919980e)
![image](https://github.com/NCAR/hrldas/assets/36647844/b5bd67ff-cbf7-4a1e-a4c6-fd73b7edda34)

**Reply by sudarsan3010** [reply]
I have executed the spinup run for 40 years using,  SPINUP_LOOPS = 40.
Now, my variables of interest like SFCRNOFF, UGDRNOFF and WT are giving the accumulated numbers for the whole 40 years. So, If I am using this as an initial condition for my simulation, the results will not match the in-situ validation data. 
Is there any way to get the output of current surface and sub-surface runoff, without the accumulated one?

**Reply by cenlinhe** [reply]
Unfortunately, the default runoff is accumulated quantity. The easiest way is to reset the SFCRNOFF, UGDRNOFF in your restart file to be 0 after spinup run (before your official run). Then, in your official run output files, use SFCRNOFF, UGDRNOFF from the next day to minus the current day to get the runoff for this day.
You may also modify the code directly to output the instantaneous runoff. We will add this later in our next code udpate.

**Reply by sudarsan3010** [reply]
Only Modifying the following file will be sufficient?! 
[noahmp](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84)/[drivers](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84/drivers)/[hrldas](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84/drivers/hrldas)
/WaterVarOutTransferMod.F90

**From:**
NoahmpIO%SFCRUNOFF   (I,J) = NoahmpIO%SFCRUNOFF(I,J) + noahmp%water%flux%RunoffSurface
NoahmpIO%UDRUNOFF    (I,J) = NoahmpIO%UDRUNOFF (I,J) + noahmp%water%flux%RunoffSubsurface
NoahmpIO%QTDRAIN     (I,J) = NoahmpIO%QTDRAIN  (I,J) + noahmp%water%flux%TileDrain
NoahmpIO%RECHXY      (I,J) = NoahmpIO%RECHXY(I,J) + (noahmp%water%state%RechargeGwShallowWT*1.0e3)
NoahmpIO%DEEPRECHXY  (I,J) = NoahmpIO%DEEPRECHXY(I,J) + noahmp%water%state%RechargeGwDeepWT

**To:**
NoahmpIO%SFCRUNOFF   (I,J) = noahmp%water%flux%RunoffSurface
NoahmpIO%UDRUNOFF    (I,J) = noahmp%water%flux%RunoffSubsurface
NoahmpIO%QTDRAIN     (I,J) = noahmp%water%flux%TileDrain
NoahmpIO%RECHXY      (I,J) = (noahmp%water%state%RechargeGwShallowWT*1.0e3)
NoahmpIO%DEEPRECHXY  (I,J) =  noahmp%water%state%RechargeGwDeepWT

**Reply by SmithSimmon** [reply]
> ![image](https://private-user-images.githubusercontent.com/36647844/345085521-a15f459c-618f-4fce-a408-e5f59919980e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0MTc5MDMsIm5iZiI6MTcyMjQxNzYwMywicGF0aCI6Ii8zNjY0Nzg0NC8zNDUwODU1MjEtYTE1ZjQ1OWMtNjE4Zi00ZmNlLWE0MDgtZTVmNTk5MTk5ODBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMxVDA5MjAwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZGVkMTY4YWM3YzVhMGQ0M2UyMDI3ZjlhNDkzYzg2YTU3ZmJmOTEzNmNkNWZmOWZmM2NiOTYxNDU2MDk5N2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QfrNe4ZQ6E8WrOc6osb12waRZhw2_PNyO76bhNZ6T5A) ![image](https://private-user-images.githubusercontent.com/36647844/345085738-b5bd67ff-cbf7-4a1e-a4c6-fd73b7edda34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0MTc5MDMsIm5iZiI6MTcyMjQxNzYwMywicGF0aCI6Ii8zNjY0Nzg0NC8zNDUwODU3MzgtYjViZDY3ZmYtY2JmNy00YTFlLWE0YzYtZmQ3M2I3ZWRkYTM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMxVDA5MjAwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1MDA4ODI4MDM5ZTRlOTVlZTc5MDIwYTlhNTdhMjBlZjhhNzY0NzdiYjZhNTQ4MjAwZDVmMjU1Y2M4ZTk5MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JuL9A4fE0Sd8jV0OOYTdr9bxAbhbAFPqz1ztHSr-80s)

Hi, @RAbolafiaRosenzweig . For the second reply, I have executed  the spinup from 1981 to 1995, totolly 15 years. I set the RESTART_FREQUENCY_HOURS=131472 (5478 days from 1981 to 1995). The time flag of restart file was 1996-01-01 00:00:00. If I want to get the initial state of 1981-01-01 00:00:00 for my official simulation, should I change the time flag of restat file to 1981-01-01 following the command line you posted? Any suggestions will be appriciated.

**Reply by RAbolafiaRosenzweig** [reply]
Yes, this can be done with the following nco command:
ncap2 -O -s 'Times(0,:)="1981-01-01_00:00:00";' RESTART.1996010100_DOMAIN1 RESTART.1981010100_DOMAIN1

**Reply by cenlinhe** [reply]
For your runoff output code modification, yes, it is the change you need to make. Note that the unit will be mm per model soil timestep instead of mm/s.

**Reply by sudarsan3010** [reply]
I appreciate your confirmation. I am sorry if I sound dumb, but as you mentioned the model timestep, I am curious to know this.

I am running my model time step at 30 min and the output interval is 1 day.

NOAH_TIMESTEP    = 1800
OUTPUT_TIMESTEP  = 86400

So the units,
SFCRUNOFF, UDRUNOFF - mm/30mins or mm/day?
RECH, DEEPRECH - m/30mins or m/day?
**********************************

Another query apart from this is when using option-5 (MMF scheme), the hrldas.exe runs in not more than 22 cores. If I assign the jobs to more cores, the model neither gives output nor error (Checking the log file, I understood it stops particularly at a line and doesn't proceed).

**Reply by cenlinhe** [reply]
it is mm/30mins and m/30mins.
THe MMF scheme parallelization part has some issue currently, and we are trying to fix it this and next week.

**Reply by SmithSimmon** [reply]
> Yes, this can be done with the following nco command: ncap2 -O -s 'Times(0,:)="1981-01-01_00:00:00";' RESTART.1996010100_DOMAIN1 RESTART.1981010100_DOMAIN1

Thanks for you reply

**Reply by sudarsan3010** [reply]
> it is mm/30mins and m/30mins. THe MMF scheme parallelization part has some issue currently, and we are trying to fix it this and next week.

Thank you. Kindly let me know the link where can I follow this update (if exists); Or just update here in the same threat once the fix is done. Thanks in advance.

**Reply by sudarsan3010** [reply]
> it is mm/30mins and m/30mins. THe MMF scheme parallelization part has some issue currently, and we are trying to fix it this and next week.
Hi @cenlinhe,

The values from the MMF scheme's recharge variable seem a little weird. I assumed my Recharge (from MMF) is in the unit of m/30min, multiplied by 48 to convert to m/day. When I computed the annual sum, these values did not fit into the zone of Insitu recharge. The Insitu data ranges from -0.5 to 0.6 whereas the MMF recharge variable ranges from -100 to 300.

![MMF_x48](https://github.com/user-attachments/assets/5e6fe2c9-210d-40fd-a6e7-36fe42c3d201)


Is it because the recharge variable in the MMF scheme is multiplied by the grid resolutions as per the below equation?
If so, my model resolution is 9kmx 9km, If I divide it by that, the value reduces to the negligible values.

![image](https://github.com/user-attachments/assets/0b1b50e4-25d2-453b-b1ca-6b43bff9af84)
____________________________________________________________________

NoahmpIO%RECHXY (I,J) = (noahmp%water%state%RechargeGwShallowWT*1.0e3

What is the purpose of **1.0e3** in the above equation?

**Note:**
As mentioned in the above comments, I am computing instantaneous recharge, instead of cumulative one
NoahmpIO%RECHXY (I,J) = (noahmp%water%state%RechargeGwShallowWT*1.0e3)

**Reply by cenlinhe** [reply]
@CharlesZheZhang Could you please take a look at this question? Thanks!

**Reply by CharlesZheZhang** [reply]
Hi @sudarsan3010 , it seems like your model time step is 1800 (30min), but the output time step is a day.
So what you get from the output is the last timestep (30min) of the output timestep.
Another point I want to make is, the RECH variable is an accumulated value, so it is 0 at the beginning of the simulation and gets larger, see this line:
https://github.com/NCAR/noahmp/blob/c7a04d8c6d4b9533a6b99c72229ed9279fe39e46/src/GroundWaterMmfMod.F90#L208

The *1.0e3 in this line is to convert unit to mm. (I checked with some WRF output to confirm RECH should be in mm)

The recharge variable is an accumulated water flux so it doesn't matter of your grid size, it's an average value for the whole grid. So, If you want to compare monthly recharge value with the observation, you may want to use the RECH at the end of the month minus at the beginning of the month to get the total accumulated recharge for this month.

**Reply by sudarsan3010** [reply]
Hi @CharlesZheZhang,

Okay. Considering recharge as mm solved the issue of value difference between in situ and modeled.

I intend to calculate the instantaneous recharge of each day not the accumulated one.
I have modified the below codes for that,
[noahmp](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84)/[drivers](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84/drivers)/[hrldas](https://github.com/NCAR/noahmp/tree/fa2d2098bac1e437a21ccc15494b152996d5bc84/drivers/hrldas)
/WaterVarOutTransferMod.F90

From:
NoahmpIO%SFCRUNOFF (I,J) = NoahmpIO%SFCRUNOFF(I,J) + noahmp%water%flux%RunoffSurface
NoahmpIO%UDRUNOFF (I,J) = NoahmpIO%UDRUNOFF (I,J) + noahmp%water%flux%RunoffSubsurface
NoahmpIO%QTDRAIN (I,J) = NoahmpIO%QTDRAIN (I,J) + noahmp%water%flux%TileDrain
NoahmpIO%RECHXY (I,J) = NoahmpIO%RECHXY(I,J) + (noahmp%water%state%RechargeGwShallowWT*1.0e3)
NoahmpIO%DEEPRECHXY (I,J) = NoahmpIO%DEEPRECHXY(I,J) + noahmp%water%state%RechargeGwDeepWT

To:
NoahmpIO%SFCRUNOFF (I,J) = noahmp%water%flux%RunoffSurface
NoahmpIO%UDRUNOFF (I,J) = noahmp%water%flux%RunoffSubsurface
NoahmpIO%QTDRAIN (I,J) = noahmp%water%flux%TileDrain
NoahmpIO%RECHXY (I,J) = (noahmp%water%state%RechargeGwShallowWT*1.0e3)
NoahmpIO%DEEPRECHXY (I,J) = noahmp%water%state%RechargeGwDeepWT

Only modifying at this file WaterVarOutTransferMod.F90 will do? or should I need to correct at the [noahmp](https://github.com/NCAR/noahmp/tree/c7a04d8c6d4b9533a6b99c72229ed9279fe39e46)/[src](https://github.com/NCAR/noahmp/tree/c7a04d8c6d4b9533a6b99c72229ed9279fe39e46/src)
/GroundWaterMmfMod.F90 also?
